packet_encoder.go

```diff
@@ -19,9 +19,15 @@ type packetEncoder interface {
 	putInt32Array(in []int32) error
 	putInt64Array(in []int64) error
 
+	// Provide the current offset to record the batch size metric
+	offset() int
+
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// To only record metrics during the real encoder pass
+	doRecordMetrics() bool
```

Comment:
Sounds good for using `metricsRegistry() metrics.Registry` (it's an interface) to the `packetEncoder` interface.

If I understand correctly, the broker would pass the `metrics.Registry` reference during encoding that would eventually get stored into the `realEncoder struct`:
```
broker.send(...) > encode(req) > {req.encode(prep_encoder); req.encode(real_encoder)}
```
Do you prefer to add a `metrics.Registry` argument to `encode(e encoder) ([]byte, error)` or add another function `encodeWithMetrics(e encoder, metricsRegistry metrics.Registry) ([]byte, error)` like in the initial 1e5e537 commit?

The other places that uses `encode(e encoder) ([]byte, error)` would then either pass `nil` or call the new `encodeWithMetrics` function.